tests/e2e/e2e_test.go

```diff
@@ -256,29 +334,406 @@ func (s *IntegrationTestSuite) TestConcentratedLiquidity() {
+
+	// Assert that the balance changed only for tokenIn (uosmo)
+	assertDefaultBalances(addr1BalancesBefore, addr1BalancesAfter, false, true)
+
+	// Assert Balances: Swap 1
+
+	// Calculate uncollected fees for address1 position1
+	feesUncollectedAddress1Position1_Swap1 := calculateUncollectedFees(
+		positionsAddress1[0].Liquidity,
+		sdk.ZeroDec(), // no growth below
+		sdk.ZeroDec(), // no growth above
+		sdk.ZeroDec(), // no feeGrowthInsideLast - it is the first swap
+		feeGrowthGlobal,
+	)
+
+	// Assert
+	s.Require().Equal(
+		addr1BalancesBefore.AmountOf("uosmo").Add(feesUncollectedAddress1Position1_Swap1.TruncateInt()),
+		addr1BalancesAfter.AmountOf("uosmo"),
+	)
+
+	// Swap 2
+	//
+	// Cross-tick swap:
+	// * Part of swap happens in range of liquidity of 3 positions: both of address1 and one for address3 (until tick 400 - upper tick of address1 position1)
+	// * Another part happens in range of liquidity of 2 positions: one from address1 and address3
+	//
+	// Asserts:
+	// * Net liquidity is kicked out when crossing initialized tick
+	// * Liquidity of position that was kicked out after first swap step does not earn rewards from second swap step
+	// * Uncollected fees from multiple swaps are correctly summed up and collected
+
+	// tickOffset is a tick index after the next initialized tick to which this swap needs to move the current price
+	tickOffset := sdk.NewInt(3)
+	sqrtPriceBeforeSwap = concentratedPool.GetCurrentSqrtPrice()
+	liquidityBeforeSwap = concentratedPool.GetLiquidity()
+	nextInitTick := sdk.NewInt(400) // address1 position1's upper tick
+
+	// Calculate sqrtPrice after and at the next initialized tick (upperTick of address1 position1 - 400)
+	// sqrtPriceAfterNextInitializedTick := calculateSqrtPriceAtTick(nextInitTick.Add(tickOffset), sdk.NewInt(precisionFactorAtPriceOne))
+	sqrtPriceAfterNextInitializedTick, err := cl.TickToSqrtPrice(nextInitTick.Add(tickOffset), sdk.NewInt(precisionFactorAtPriceOne))
+	s.Require().NoError(err)
+	sqrtPriceAtNextInitializedTick, err := cl.TickToSqrtPrice(nextInitTick, sdk.NewInt(precisionFactorAtPriceOne))
+	s.Require().NoError(err)
+
+	// Calculate Δ(sqrtPrice):
+	// deltaSqrtPriceAfterNextInitializedTick = ΔsqrtP(403) - ΔsqrtP(400)
+	// deltaSqrtPriceAtNextInitializedTick = ΔsqrtP(400) - ΔsqrtP(currentTick)
+	deltaSqrtPriceAfterNextInitializedTick := sqrtPriceAfterNextInitializedTick.Sub(sqrtPriceAtNextInitializedTick)
+	deltaSqrtPriceAtNextInitializedTick := sqrtPriceAtNextInitializedTick.Sub(sqrtPriceBeforeSwap)
+
+	// Calculate the amount of osmo required to:
+	// * amountInToGetToTickAfterInitialized - move price from next initialized tick (400) to destination tick (400 + tickOffset)
+	// * amountInToGetToNextInitTick - move price from current tick to next initialized tick
+	// Formula is as follows:
+	// Δy = L * Δ(sqrtPrice)
+	amountInToGetToTickAfterInitialized := deltaSqrtPriceAfterNextInitializedTick.Mul(liquidityBeforeSwap.Sub(positionsAddress1[0].Liquidity))
```

Comment:
I think we should make all these calculations into functions, seeing that we are repeatedly doing this calculation + would have us with way better readability

Reply:
It can be done, but only partially:

Swap2 and Swap3 use different formulas to calculate amounts needed:

Swap2 uses: `CalcAmount0Delta = liquidity * (sqrtPriceB - sqrtPriceA)) / (sqrtPriceB * sqrtPriceA`
Swap3 uses: `CalcAmount1Delta = liquidity * (sqrtPriceB - sqrtPriceA)`

I can make a function to calculate `liquidity * (sqrtPriceB - sqrtPriceA)` to use in both of these calculations, but do we want to do that?